In [1]:
import pandas as pd
import plotly.graph_objects as go
import ast

DATA_FOLDER = './data/Cleantech Media Dataset'

In [2]:
df = pd.read_csv(f'{DATA_FOLDER}/cleantech_media_dataset_v2_2024-02-23.csv')

In [3]:
df.head()

Unnamed: 0                                              title        date  \
0        1280  Qatar to Slash Emissions as LNG Expansion Adva...  2021-01-13   
1        1281               India Launches Its First 700 MW PHWR  2021-01-15   
2        1283              New Chapter for US-China Energy Trade  2021-01-20   
3        1284  Japan: Slow Restarts Cast Doubt on 2030 Energy...  2021-01-22   
4        1285     NYC Pension Funds to Divest Fossil Fuel Shares  2021-01-25   

  author                                            content       domain  \
0    NaN  ["Qatar Petroleum ( QP) is targeting aggressiv...  energyintel   
1    NaN  ["• Nuclear Power Corp. of India Ltd. ( NPCIL)...  energyintel   
2    NaN  ["New US President Joe Biden took office this ...  energyintel   
3    NaN  ["The slow pace of Japanese reactor restarts c...  energyintel   
4    NaN  ["Two of New York City's largest pension funds...  energyintel   

                                                 url  
0  https://www.energyintel.com/0000017b-a7dc-de4c...  
1  https://www.energyintel.com/0000017b-a7dc-de4c...  
2  https://www.energyintel.com/0000017b-a7dc-de4c...  
3  https://www.energyintel.com/0000017b-a7dc-de4c...  
4  https://www.energyintel.com/0000017b-a7dc-de4c...

In [4]:
columns = df.columns
total_counts = df.count()
nan_counts = df.isna().sum()
unique_counts = df.nunique()

nan_percentages = (nan_counts / len(df)) * 100
unique_percentages = (unique_counts / len(df)) * 100

data = {
    'Total Count': total_counts,
    'NaN Count': nan_counts,
    'NaN Percentage (%)': nan_percentages,
    'Unique Count': unique_counts,
    'Unique Percentage (%)': unique_percentages
}
summary_df = pd.DataFrame(data, index=columns)

summary_df

Total Count  NaN Count  NaN Percentage (%)  Unique Count  \
Unnamed: 0         9593          0            0.000000          9593   
title              9593          0            0.000000          9569   
date               9593          0            0.000000           967   
author               31       9562           99.676848             7   
content            9593          0            0.000000          9588   
domain             9593          0            0.000000            19   
url                9593          0            0.000000          9593   

            Unique Percentage (%)  
Unnamed: 0             100.000000  
title                   99.749818  
date                    10.080267  
author                   0.072970  
content                 99.947879  
domain                   0.198061  
url                    100.000000

In [5]:
domain_freq = df['domain'].value_counts()
domain_freq = domain_freq.reset_index()
domain_freq.columns = ['domain', 'count']

fig = go.Figure()
fig.add_trace(go.Bar(x=domain_freq['domain'], y=domain_freq['count']))

fig.update_layout(
    title='Frequency of Publishers in Cleantech',
    xaxis_title='Domain',
    yaxis_title='Frequency'
)

fig.show()

### Taking a closer look at titles
As the summary has shown, only `9569` of the `9593` scraped resources in the dataset have a unique title. This subsection explores if these "duplicate titles" have an underlying error or if these occurences of duplicates can be ignored.

In [6]:
title_freq = df['title'].value_counts()
title_freq = title_freq[title_freq > 1]
title_freq = title_freq.reset_index()
title_freq.columns = ['title', 'count']

title_freq

title  count
0                    Cleantech Thought Leaders Series      5
1                                About David J. Cross      5
2             Cleantech Insights from Industry Series      4
3   Truss, Johnson Join Rebellion Against Sunak fo...      2
4   Staggering potential in next-gen building bloc...      2
5   Royal Opera House drops BP as sponsor after 33...      2
6   Macquarie targets North Sea as the green energ...      2
7   BEIS mulls ringfenced CfD support for geotherm...      2
8   Five ways for the energy industry to shift the...      2
9   Key trends in UK renewables and what to expect...      2
10  Green hydrogen seen competing with LNG within ...      2
11  Aberdeen’ s NZTC plans national centre for geo...      2
12  Liz Truss opposes solar panels on farmland, Do...      2
13  XR goes big on fake oil in protests at SLB, In...      2
14                           Top Climate Stories 2021      2
15  GE blocked from selling huge offshore turbine ...      2

Now, lets take a closer look at the contents of the suspected duplicate documents.

In [7]:
def calculate_all_duplicate_document_contents(df, title_freq):
    duplicates_counts = {}
    
    for title in title_freq['title']:
        duplicate_contents = df[df['title'] == title]['content']
        duplicate_contents = duplicate_contents.apply(ast.literal_eval)
        duplicate_contents = duplicate_contents.explode()
        duplicates_count = duplicate_contents.duplicated().sum()
        duplicates_counts[title] = duplicates_count
    
    # return a pandas dataframe with the title and the count of duplicated contents
    return pd.DataFrame(list(duplicates_counts.items()), columns=['title', 'duplicated_count'])

duplicated_title_contents = calculate_all_duplicate_document_contents(df, title_freq)

duplicated_title_contents

title  duplicated_count
0                    Cleantech Thought Leaders Series                30
1                                About David J. Cross                48
2             Cleantech Insights from Industry Series                21
3   Truss, Johnson Join Rebellion Against Sunak fo...                 9
4   Staggering potential in next-gen building bloc...                39
5   Royal Opera House drops BP as sponsor after 33...                10
6   Macquarie targets North Sea as the green energ...                12
7   BEIS mulls ringfenced CfD support for geotherm...                17
8   Five ways for the energy industry to shift the...                15
9   Key trends in UK renewables and what to expect...                10
10  Green hydrogen seen competing with LNG within ...                 4
11  Aberdeen’ s NZTC plans national centre for geo...                15
12  Liz Truss opposes solar panels on farmland, Do...                13
13  XR goes big on fake oil in protests at SLB, In...                12
14                           Top Climate Stories 2021                 3
15  GE blocked from selling huge offshore turbine ...                 6

The function yielded that the duplicate observations of title contain actual duplicate information on chunk-basis.

This could mean that there are even more duplicate chunks under titles that aren't duplicate, so lets next look at that:

In [8]:
df['content'] = df['content'].apply(ast.literal_eval)
df_exploded_contents = df.explode('content')

In [9]:
print(f'Total duplicated contents: {df_exploded_contents.duplicated().sum()}'
      f'\nTotal duplicated contents from duplicated titles: {duplicated_title_contents["duplicated_count"].sum()}'
      f'\nTotal duplicated contents from non-duplicated titles: {df_exploded_contents.duplicated().sum()-duplicated_title_contents["duplicated_count"].sum()}')

Total duplicated contents: 893
Total duplicated contents from duplicated titles: 264
Total duplicated contents from non-duplicated titles: 629


As the result shows, another `629` chunks on top of the `264` duplicates inside the duplicate-title-occurences emerged.

### Analyzing Languages

In [13]:
df

Unnamed: 0                                              title  \
0           1280  Qatar to Slash Emissions as LNG Expansion Adva...   
1           1281               India Launches Its First 700 MW PHWR   
2           1283              New Chapter for US-China Energy Trade   
3           1284  Japan: Slow Restarts Cast Doubt on 2030 Energy...   
4           1285     NYC Pension Funds to Divest Fossil Fuel Shares   
...          ...                                                ...   
9588       81812  Strata Clean Energy Nets $ 300 Million in Fund...   
9589       81813  Orsted Deploying SparkCognition Renewable Suit...   
9590       81814     Veolia Has Plans for 5 MW of Solar in Arkansas   
9591       81815                      SunEdison: Too Big, Too Fast?   
9592       81816  Vikings Solar-Plus-Storage Development Nets Fi...   

            date author                                            content  \
0     2021-01-13    NaN  [Qatar Petroleum ( QP) is targeting aggressive...   
1     2021-01-15    NaN  [• Nuclear Power Corp. of India Ltd. ( NPCIL) ...   
2     2021-01-20    NaN  [New US President Joe Biden took office this w...   
3     2021-01-22    NaN  [The slow pace of Japanese reactor restarts co...   
4     2021-01-25    NaN  [Two of New York City's largest pension funds ...   
...          ...    ...                                                ...   
9588  2023-11-06    NaN  [Strata Clean Energy has closed a $ 300 millio...   
9589  2023-11-07    NaN  [Global renewable energy developer Ørsted is d...   
9590  2023-11-07    NaN  [Veolia North America, a provider of environme...   
9591  2023-11-08    NaN  [Once the self-proclaimed “ leading renewable ...   
9592  2023-11-08    NaN  [Arevon Energy Inc. has closed financing on th...   

                domain                                                url  
0          energyintel  https://www.energyintel.com/0000017b-a7dc-de4c...  
1          energyintel  https://www.energyintel.com/0000017b-a7dc-de4c...  
2          energyintel  https://www.energyintel.com/0000017b-a7dc-de4c...  
3          energyintel  https://www.energyintel.com/0000017b-a7dc-de4c...  
4          energyintel  https://www.energyintel.com/0000017b-a7dc-de4c...  
...                ...                                                ...  
9588  solarindustrymag  https://solarindustrymag.com/strata-clean-ener...  
9589  solarindustrymag  https://solarindustrymag.com/orsted-deploying-...  
9590  solarindustrymag  https://solarindustrymag.com/veolia-has-plans-...  
9591  solarindustrymag  http://www.solarindustrymag.com/online/issues/...  
9592  solarindustrymag  https://solarindustrymag.com/vikings-solar-plu...  

[9593 rows x 7 columns]

In [14]:
from langdetect import detect, LangDetectException
import numpy as np

def safe_detect(text):
    text = str(text)
    try:
        return detect(text)
    except LangDetectException:
        return np.nan

df['language'] = df['content'].apply(safe_detect)

In [17]:
df['language'].value_counts()

language
en    9589
de       3
ru       1
Name: count, dtype: int64